bi/pkg/specs/wait.go

```diff
@@ -138,25 +145,30 @@ func batteryInfoConfigMapWatchOpts(ns string) *kube.WatchOptions {
 				},
 			},
 		})},
-		Callback: func(u *unstructured.Unstructured) (bool, error) {
-			var cm corev1.ConfigMap
-			err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &cm)
-			if err != nil {
-				slog.Debug(
-					"failed to convert unstructured object into typed resource",
-					slog.String("namespace", u.GetNamespace()),
-					slog.String("name", u.GetName()),
-					slog.Any("err", err),
-				)
-				return false, nil
-			}
-
+		Callback: buildCallback(func(cm *corev1.ConfigMap) (bool, error) {
 			// keep watching as long as hostname isn't set
 			if cm.Data["hostname"] != "" {
 				slog.Debug("got control server hostname", slog.String("hostname", cm.Data["hostname"]))
 				return true, nil
 			}
 			return false, nil
-		},
+		}),
+	}
+}
+
+func buildCallback[T any](fn testFn[T]) func(u *unstructured.Unstructured) (bool, error) {
+	return func(u *unstructured.Unstructured) (bool, error) {
+		var obj *T
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj)
+		if err != nil {
+			slog.Debug(
+				"failed to convert unstructured object into typed resource",
+				slog.String("namespace", u.GetNamespace()),
+				slog.String("name", u.GetName()),
+				slog.Any("err", err),
+			)
+			return false, nil
+		}
+		return fn(obj)
```

Comment:
This same pattern was repeated for each of the watches. Refactored and made it generic to keep things D.R.Y.